### PR TITLE
Revise erf OP

### PR DESCRIPTION
**Pull request by luweizhou2016**
### Details:
 - *Revise Erf OP spec and OP implement*
 - *Add Erf OP visitor test and type_prop test *
 - *Migrate existing tanh and mish to use templeate code*
 - *Migrate tanh OP to use RTTI to fix unit-test issue*

### Tickets:
 - *37437*

### CODE REVIEW

File: docs/ops/arithmetic/Erf_1.md [hunk 1]
```diff
@@ -13,17 +13,15 @@ For each element from the input tensor calculates corresponding element in the o
...
 
-**Attributes**:
-
-    No attributes available.
+**Attributes**: *Erf* operation has no attributes.
 
 **Inputs**
 
-* **1**: A tensor of type T. **Required.**
+* **1**: A tensor `x` of type `T` and arbitrary shape. **Required**.
```
Comment (ilyachur) [inline docs/ops/arithmetic/Erf_1.md:20]: As I see we have a different agreement for highlighting  `T`.
You can find it here: https://github.com/openvinotoolkit/openvino/pull/5618/

@lazarevevgeny Please have a look.

**Comment by ggalieroc** [comment]
I don't agree to introduce changes of a different operation than erf in this PR. It is confusing for reviewers and in case of regression it makes difficult to track as these commits are eventually squashed into one. Please revert these changes and address them in a separate PR if needed

File: docs/ops/arithmetic/Erf_1.md (hunk 1 above)
Comment (pelszkow) [reply docs/ops/arithmetic/Erf_1.md:20]: Who we should mark `required` (where put the dot):
1) `**Required.**`
2) `**Required**.`
in docs/ops/ more often we use 1st approach 180/38

Comment (lazarevevgeny) [reply docs/ops/arithmetic/Erf_1.md:20]: We should wrap types with `*` characters. 

Regarding the "Required". I personally like the `**Required**.` approach, but we can ask @openvinotoolkit/openvino-docs-maintainers  on what is the better option.

Comment (luweizhou2016) [reply docs/ops/arithmetic/Erf_1.md]: Both option 1, 2 fromat exist in current doc code,  I will choose the majority.

**Comment by luweizhou2016** [comment]
> I don't agree to introduce changes of a different operation than erf in this PR. It is confusing for reviewers and in case of regression it makes difficult to track as these commits are eventually squashed into one. Please revert these changes and address them in a separate PR if needed

Okay. Should I create new JIRA or reopen the previous JIRA?

**Comment by ilyachur** [comment]
> > I don't agree to introduce changes of a different operation than erf in this PR. It is confusing for reviewers and in case of regression it makes difficult to track as these commits are eventually squashed into one. Please revert these changes and address them in a separate PR if needed
> 
> Okay. Should I create new JIRA or reopen the previous JIRA?

Will be better to have a separate ticket.

File: inference-engine/tests/functional/inference_engine/serialization/single_layer/activation.cpp [hunk 2]
```diff
@@ -57,7 +57,6 @@ const std::map<ActivationTypes, std::vector<std::vector<float>>> activationTypes
         {HSigmoid,              {}},
         {RoundHalfToEven,       {}},
         {RoundHalfAwayFromZero, {}},
-        {Erf,                   {}},
```
Comment (ilyachur) [inline inference-engine/tests/functional/inference_engine/serialization/single_layer/activation.cpp:60]: Why did you remove `Erf` serialization tests?

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/activation.cpp [hunk 3]
```diff
@@ -57,7 +57,6 @@ const std::map<ActivationTypes, std::vector<std::vector<float>>> activationTypes
         {HSigmoid,              {}},
         {RoundHalfToEven,       {}},
         {RoundHalfAwayFromZero, {}},
-        {Erf,                   {}},
```
Comment (ilyachur) [inline inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/activation.cpp:60]: Why did you remove `Erf` from SLT?

File: inference-engine/tests/functional/inference_engine/serialization/single_layer/activation.cpp (hunk 2 above)
Comment (luweizhou2016) [reply inference-engine/tests/functional/inference_engine/serialization/single_layer/activation.cpp:60]: It is duplicated . See the line 50 in the file.

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/activation.cpp (hunk 3 above)
Comment (luweizhou2016) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/activation.cpp:60]: Also duplicated  element.

File: docs/ops/arithmetic/Erf_1.md [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
 **Category**: Arithmetic unary operation
 
-**Short description**: *Erf* calculates the Gauss error function element-wise with given tensor.
+**Short description**:  *Erf* performs element-wise Gauss error function (erf) with given tensor.
```
Comment (ggalieroc) [inline docs/ops/arithmetic/Erf_1.md:7]: nitpick: remove double space before \*Erf\*

File: docs/ops/arithmetic/Erf_1.md [hunk 5]
```diff
@@ -13,17 +13,15 @@ For each element from the input tensor calculates corresponding element in the o
...
 
-**Attributes**:
-
-    No attributes available.
+**Attributes**: *Erf* operation has no attributes.
 
 **Inputs**
 
-* **1**: A tensor of type *T*. **Required.**
+* **1**: A tensor `x` of type *T* and arbitrary shape. **Required.**
```
Comment (ggalieroc) [inline docs/ops/arithmetic/Erf_1.md:20]: nitpick: In general when there is only one input we do not label it with a name (for simplicity). We refer to it as input tensor. You can apply this to Output section too

File: docs/ops/arithmetic/Erf_1.md [hunk 6]
```diff
@@ -13,17 +13,15 @@ For each element from the input tensor calculates corresponding element in the o
...
 
 **Outputs**
 
-* **1**: The result of element-wise operation. A tensor of type *T*.
+* **1**: The result of element-wise *Erf* function applied to the input tensor `x`. A tensor of type *T* and the same shape as `x` input tensor.
 
 **Types**
```
Comment (ggalieroc) [inline docs/ops/arithmetic/Erf_1.md:27]: floating point type ---> floating-point

**Review by ggalieroc — COMMENTED** [review]
nitpick: Please also consider to address this [comment ](https://github.com/openvinotoolkit/openvino/pull/6288#discussion_r660644993)for consistency with other specs

**Comment by luweizhou2016** [comment]
Thanks and updated.

**Comment by ggalieroc** [comment]
Please add this operation (Erf-1) to the [list of trusted operators](https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/tests/ie_test_utils/functional_test_utils/layer_tests_summary/utils/constants.py). This is required for ref. implementations.
Other than that, LGTM. Approving beforehand

**Review by ggalieroc — APPROVED** [review]

**Comment by luweizhou2016** [comment]
> Please add this operation (Erf-1) to the [list of trusted operators](https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/tests/ie_test_utils/functional_test_utils/layer_tests_summary/utils/constants.py). This is required for ref. implementations.
> Other than that, LGTM. Approving beforehand

Done

**Comment by luweizhou2016** [comment]
@ilyachur , Can you pls help to review/merge? Thanks!

**Comment by ilyachur** [comment]
@rkazants Could you review the specification?

File: ngraph/test/backend/erf.in.cpp [hunk 7]
```diff
@@ -31,22 +31,38 @@ using namespace ngraph;
...
+
+NGRAPH_TEST(${BACKEND_NAME}, erf_4D)
+{
+    Shape shape{2, 2, 2, 1};
+    auto A = make_shared<op::Parameter>(element::f32, shape);
+    auto f = make_shared<Function>(make_shared<op::Erf>(A), ParameterVector{A});
+
+    auto test_case = test::TestCase<TestEngine>(f);
+    vector<float> in_vec{-8.8f, -5.4f, -3.1f, -2.5f, 0, 1.7f, 3.2f, 4.9f};
+    vector<float> out_vec{in_vec};
```
Comment (rkazants) [inline ngraph/test/backend/erf.in.cpp:60]: also, I recommend to check for both (+/-)inf values in input.

Comment (luweizhou2016) [reply ngraph/test/backend/erf.in.cpp]: checked in the latest commit

**Comment by luweizhou2016** [comment]
@rkazants @ilyachur Pls help to review changes.

File: docs/template_plugin/tests/functional/op_reference/erf.cpp [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        return std::make_shared<Function>(NodeVector {erf}, ParameterVector {in});
+    }
+};
+
+TEST_P(ReferenceErfLayerTest, CompareWithRefs) {
+    Exec();
+}
+
+INSTANTIATE_TEST_SUITE_P(smoke_Erf_With_Hardcoded_Refs, ReferenceErfLayerTest,
+                         ::testing::Values(ErfParams(ngraph::PartialShape {2, 5}, ngraph::element::f32,
```
Comment (ilyachur) [inline docs/template_plugin/tests/functional/op_reference/erf.cpp:67]: Could you cover all supported data types?

File: docs/template_plugin/tests/functional/op_reference/erf.cpp [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+#include <tuple>
+
+#include "base_reference_test.hpp"
+
+using namespace ngraph;
+using namespace InferenceEngine;
+
+struct ErfParams {
+    template <class IT>
+    ErfParams(const ngraph::PartialShape& shape, const ngraph::element::Type& iType, const std::vector<IT>& iValues, size_t iSize = 0, size_t oSize = 0)
```
Comment (ilyachur) [inline docs/template_plugin/tests/functional/op_reference/erf.cpp:21]: Looks like you don't use `iSize` and `oSize`

File: ngraph/test/backend/erf.in.cpp [hunk 10]
```diff
@@ -1,52 +0,0 @@
-// Copyright (C) 2018-2021 Intel Corporation
-// SPDX-License-Identifier: Apache-2.0
-//
-
```
Comment (rkazants) [inline ngraph/test/backend/erf.in.cpp:4]: why is this test removed? it is useful for different backend implementations.

Comment (luweizhou2016) [reply ngraph/test/backend/erf.in.cpp:4]: Just forwarded you the email. I think template plugin would be used to validate the reference OP.  CPU plugin will use SLT to validate OP tests.

File: docs/template_plugin/tests/functional/op_reference/erf.cpp (hunk 8 above)
Comment (luweizhou2016) [reply docs/template_plugin/tests/functional/op_reference/erf.cpp]: Add all the supported test.

File: docs/template_plugin/tests/functional/op_reference/erf.cpp [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+#include <tuple>
+
+#include "base_reference_test.hpp"
+
+using namespace ngraph;
+using namespace InferenceEngine;
+
+struct ErfParams {
+    template <class IT>
+    ErfParams(const ngraph::PartialShape& shape, const ngraph::element::Type& iType, const std::vector<IT>& iValues)
```
Comment (ilyachur) [inline docs/template_plugin/tests/functional/op_reference/erf.cpp:21]: Please rebase your branch to the latest master. TestCommon class was added to separate namespace

Comment (luweizhou2016) [reply docs/template_plugin/tests/functional/op_reference/erf.cpp:22]: Updated. Thanks!

**Review by ilyachur — APPROVED** [review]
